src/utils.py

```diff
@@ -5,32 +5,46 @@
 
 
 class WordListManager:
-    def __init__(self, word_list=None, save_on_change=True):
+    def __init__(self, word_list: list[str] | None = None, save_on_change: bool = True) -> None:
         if word_list is None:
+            # For production: work directly with the original words.word_list
             self.word_list = words.word_list
+            self.is_test_mode = False
         else:
+            # For testing: work with provided test list
             self.word_list = word_list
+            self.is_test_mode = True
         self.save_on_change = save_on_change
         self._initial_state = None
 
     def __enter__(self):
         """Context manager entry - save initial state"""
-        self._initial_state = self.word_list.copy()
+        if not self.is_test_mode:
+            self._initial_state = words.word_list.copy()
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         """Context manager exit - save changes if any were made"""
-        if self.save_on_change and self._initial_state != self.word_list:
-            self.save_to_file()
-            print("Changes saved automatically")
-        return False  # Don't suppress exceptions
+        if self.save_on_change and not self.is_test_mode:
+            if self._initial_state != words.word_list:
+                self.save_to_file()
+                print("Changes saved automatically")
+        return False
+
+    @classmethod
+    def for_testing(cls, test_word_list=None):
```

Comment:
we can remove this method, tests can just set up state like this: `mgr = WordManager(word_list=["foo", "bar"], save_on_change=False)`